src/scf/newton.jl

```diff
@@ -0,0 +1,317 @@
+        Mstr = collinear ? "   $((@sprintf "%6.3f" round(magn, sigdigits=4))[1:6])" : ""
+        @printf "% 3d   %s   %s   %2.2e%s\n" info.n_iter Estr ΔE Δρ Mstr
+        prev_energies = info.energies
+
+        flush(stdout)
+        info
+    end
+end
+
+"""
+    solve_ΩplusK(basis::PlaneWaveBasis{T}, ψ, res, occupation;
+                 tol_cg=1e-10, verbose=false) where T
+
+Return δψ where (Ω+K) δψ = rhs
+"""
+function solve_ΩplusK(basis::PlaneWaveBasis{T}, ψ, rhs, occupation;
+                     tol_cg=1e-10, verbose=false) where T
+
+    Nk = length(basis.kpoints)
+
+    # compute quantites at the point which define the tangent space
+    ρ = compute_density(basis, ψ, occupation)
+    _, H = energy_hamiltonian(basis, ψ, occupation; ρ=ρ)
+
+    # pack and unpack
+    pack(ψ) = pack_ψ(basis, ψ)
+    unpack(x) = unpack_ψ(basis, x)
+
+    # project rhs on the tangent space before starting
+    proj_tangent!(rhs, ψ)
+    rhs_pack = pack(rhs)
+
+    # preconditioner
+    Pks = [PreconditionerTPA(basis, kpt) for kpt in basis.kpoints]
+    for ik = 1:length(Pks)
+        precondprep!(Pks[ik], ψ[ik])
+    end
+    function f_ldiv!(x, y)
+        δψ = unpack(y)
+        proj_tangent!(δψ, ψ)
+        Pδψ = [ Pks[ik] \ δψk for (ik, δψk) in enumerate(δψ)]
+        proj_tangent!(Pδψ, ψ)
+        x .= pack(Pδψ)
+    end
+
+    # Rayleigh-coefficients
+    Λ = map(enumerate(ψ)) do (ik, ψk)
+        Hk = H.blocks[ik]
+        Hψk = Hk * ψk
+        ψk'Hψk
+    end
+
+    # mapping of the linear system on the tangent space
+    function ΩpK(x)
+        δψ = unpack(x)
+        Kδψ = apply_K(basis, δψ, ψ, ρ, occupation)
+        Ωδψ = apply_Ω(δψ, ψ, H, Λ)
+        pack(Ωδψ + Kδψ)
+    end
+    J = LinearMap{T}(ΩpK, size(rhs_pack, 1))
+
+    # solve (Ω+K) δψ = rhs on the tangent space with CG
+    δψ = cg(J, rhs_pack, Pl=FunctionPreconditioner(f_ldiv!),
+            reltol=tol_cg/norm(rhs), verbose=verbose)
+
+    unpack(δψ)
+end
+
+"""
+    newton(basis::PlaneWaveBasis{T}; ψ0=nothing,
+           tol=1e-6, tol_cg=1e-10, maxiter=20, verbose=false,
+           callback=NewtonDefaultCallback(),
+           is_converged=NewtonConvergenceDensity(tol))
+
+Newton algorithm. Be careful that the starting point needs to be not too far
+from the solution.
+"""
+function newton(basis::PlaneWaveBasis{T}, ψ0;
+                tol=1e-6, tol_cg=1e-10, maxiter=20, verbose=false,
+                callback=NewtonDefaultCallback(),
+                is_converged=NewtonConvergenceDensity(tol)) where T
+
+    # setting parameters
+    model = basis.model
+    @assert model.spin_polarization in (:none, :spinless, :collinear)
+    @assert model.temperature == 0 # temperature is not yet supported
+    filled_occ = filled_occupation(model)
+    n_spin = basis.model.n_spin_components
+    n_bands = div(div(model.n_electrons, filled_occ), n_spin)
+
+    # number of kpoints and occupation
+    Nk = length(basis.kpoints)
+    occupation = [filled_occ * ones(T, n_bands) for ik = 1:Nk]
```

Comment:
Why not just `for kpt in basis.kpoints`

Reply:
Cause I reuse `Nk` so we might as well use there here too ?

Reply:
We usually don't define Nk, but fair enough.